src/hardware/pwm.rs

```diff
@@ -0,0 +1,194 @@
+///! Thermostat TEC driver IC voltage/current limits PWM driver.
+///!
+///! The Thermostat TEC driver ICs feature current limits controlled by an analog voltage input.
+///! This voltage is controlled by the MCU using low-pass filtered PWM outputs.
+///!
+use super::{
+    hal::{
+        gpio::{gpiob::*, gpioc::*, gpiod::*, gpioe::*, Alternate, AF1, AF2},
+        hal::PwmPin,
+        prelude::*,
+        pwm::{ActiveHigh, ComplementaryDisabled, ComplementaryImpossible, C1, C2, C3, C4},
+        rcc::{rec, CoreClocks},
+        stm32::{TIM1, TIM3, TIM4},
+        time::KiloHertz,
+    },
+    Channel, R_SENSE, V_PWM,
+};
+
+// PWM freqency. 20kHz is ~80dB down with the installed second order 160Hz lowpass.
+const F_PWM: KiloHertz = KiloHertz(20);
+
+/// Convert maximum current to relative pulsewidth for the (analog voltage)
+/// max output current inputs of the TEC driver.
+pub fn i_to_pwm(i: f32) -> f32 {
+    i * ((10.0 * R_SENSE) / V_PWM)
+}
+
+/// Convert maximum voltage to relative pulsewidth for the (analog voltage)
+/// max output voltage of the TEC driver.
+pub fn v_to_pwm(v: f32) -> f32 {
+    v / (4.0 * V_PWM)
+}
+
+/// TEC limit types
+///
+/// LimV - Upper and lower voltage limit
+/// LimIUp - Upper current limit
+/// LimILow - Lower current limit
+pub enum Limit {
+    LimV,
+    LimIUp,
+    LimILow,
+}
```

Comment:
Let's make the names readable. Please fewer abbreviations. No need to repeat the enum name in the enum member name.

```suggestion
pub enum Limit {
    Voltage,
    PositiveCurrent,
    NegativeCurrent,
}
```